File: designpatterns/java/src/singleton/ChocolateBoiler.java
package singleton;

public class ChocolateBoiler {
    private boolean empty;
    private boolean boiled;

    private static volatile ChocolateBoiler uniqueInstance;

    private ChocolateBoiler() {
        empty = true;
        boiled = false;
    }

    public static ChocolateBoiler getInstance() {
        if(uniqueInstance == null) {
            synchronized (ChocolateBoiler.class) {
                if(uniqueInstance == null) {
                    uniqueInstance = new ChocolateBoiler();
                }

            }
        }
        return uniqueInstance;
    }

    public void fill() {
        if(isEmpty()) {
            empty = false;
            boiled = false;
        }
    }

    public void drain() {
        if(!isEmpty() && isBoiled()) {
            empty = true;
        }
    }

    public void boil() {
        if(!isEmpty() && !isBoiled()) {
            boiled = true;
        }
    }

    public boolean isEmpty() {
        return empty;
    }

    public boolean isBoiled() {
        return boiled;
    }
}

File: designpatterns/kotlin/designpatterns/src/main/kotlin/singleton/problem/ChocolateBoiler.kt
package singleton.problem

class ChocolateBoiler private constructor(
    private var empty: Boolean = true,
    private var boiled: Boolean = false,
) {

    companion object {
        private var uniqueInstance: ChocolateBoiler? = null

        fun getInstance(): ChocolateBoiler {
            return uniqueInstance ?: ChocolateBoiler()
        }
    }

    fun fill() {
        if (isEmpty()) {
            empty = false
            boiled = false
        }
    }

    fun drain() {
        if (!isEmpty() && isEmpty()) {
            empty = true
        }
    }

    fun boil() {
        if (!isEmpty() && !isBoiled()) {
            boiled = true
        }
    }

    fun isEmpty(): Boolean {
        return empty
    }

    fun isBoiled(): Boolean {
        return boiled
    }
}